### PR TITLE
Update ctr_model.py

### PR DESCRIPTION
**Pull request by amardavid-cmd**

### CODE REVIEW

File: ctr_model.py [hunk 1]
```diff
@@ -1 +1,156 @@
...
+temp[temp > 0].plot(kind='barh')
+plt.title('Normalized missing data count per column')
+plt.xlim(0, 1)
+
+numeric_columns = ['position', 'gig_price', 'gig_avg_rating', 'gig_rated_orders', 'cnt_helpful_reviews_last_year', 'avg_review_length_last_3_months',
+       'avg_position_shown_last_60d', 'previous_order_amount']
+dates_columns = ['created_at', 'user_reg_date', 'previous_order_date']
+binary_columns = ['is_click', 'is_filtered', 'is_user_buyer','is_seller_onlie']
+categorical_columns = list(set(full_df.columns) - set(numeric_columns) - set(dates_columns) - set(binary_columns))
+
```
Comment (amardavid-cmd) [inline ctr_model.py:25]: need to add before it:
full_df[''user_timezone']=full_df[''user_timezone'].astype(str)

File: ctr_model.py [hunk 2]
```diff
@@ -1 +1,156 @@
...
+)
+temp['sc_id_ctr'] = temp.num_clicks / temp.group_size
+full_df = full_df.join(temp, on='gig_sc_id').drop(['num_clicks','group_size'], axis=1)
+
+for col in numeric_columns:
+    bins=1000
+    full_df[col].plot(kind='hist', bins=bins, figsize=(13, 5), color='b', alpha=0.5, density=True)
+    plt.title(f'{col} histogram')
+    plt.show()
+
```
Comment (amardavid-cmd) [inline ctr_model.py:42]: this I can do without for:
df[numeric_columns].hist(bins=1000)
plt.show()

File: ctr_model.py [hunk 3]
```diff
@@ -1 +1,156 @@
...
+    df['user_reg_date'] = pd.to_datetime(df['user_reg_date'])
+    df['previous_order_date'] = pd.to_datetime(df['previous_order_date'])
+
+    df['hour'] = (df.created_at.dt.hour + (df.created_at.dt.minute / 60)).round(1)
+    df['day'] = df.created_at.dt.day_name().apply(lambda x: x.lower()[:3])
+    df['week_in_month'] = df.created_at.dt.day // 7
+    df['months_since_reg'] = ((df.created_at - df.user_reg_date).dt.days / 30.2).round(1)
+    df['hour_diff_from_last_purchase'] = (df.created_at.dt.hour - df.previous_order_date.dt.hour).round(1)
+    df['days_since_last_purchase'] = (df.created_at - df.previous_order_date).dt.days
+    df['is_weekend'] = df.apply(is_weekend, axis=1)
```
Comment (amardavid-cmd) [inline ctr_model.py:61]: df['is_weekend'] = df['day'].apply(lambda x:is_weekend(x))

File: ctr_model.py [hunk 4]
```diff
@@ -1 +1,156 @@
...
+    plt.title(f'{col} histogram')
+    plt.show()
+
+layout = np.arange(np.array(binary_columns).shape[0]).reshape(-1, 2)
+fig, axs = plt.subplot_mosaic(layout, figsize=(10, 10))
+for i, col in enumerate(binary_columns):
+    ser = full_df[col].value_counts()
+    ser.plot(kind='pie', autopct='%1.1f%%', title=col, ylabel='', ax=axs[i])
+plt.show();
+
```
Comment (amardavid-cmd) [inline ctr_model.py:49]: add function 
def is_weekend(dd):
    return 1 if dd in ['sat','sun'] else 0

File: ctr_model.py [hunk 5]
```diff
@@ -1 +1,156 @@
...
+full_df = add_time_features(full_df)
+
+def rare_cats_to_other(ser, thresh=0.05):
+    value_counts = ser.value_counts(normalize=True)
+    values_to_replace = value_counts[value_counts < thresh].index
+    return ser.replace(values_to_replace, 'other')
+
+for col in categorical_columns:
+    full_df[col] = rare_cats_to_other(full_df[col])
+    full_df[col].fillna('other', inplace=True)
```
Comment (amardavid-cmd) [inline ctr_model.py:74]: check option to do it with apply on columns

Comment (amardavid-cmd) [reply ctr_model.py:74]: do it with apply on columns axis=0

File: ctr_model.py [hunk 6]
```diff
@@ -1 +1,156 @@
...
+
+for col in numeric_columns:
+    full_df[col].fillna(0, inplace=True)
```
Comment (amardavid-cmd) [inline ctr_model.py:77]: full_df[numeric_columns]=full_df[numeric_columns].fillna(0)

File: ctr_model.py [hunk 7]
```diff
@@ -1 +1,156 @@
...
+
+def compare_sc_id(row):
+    if pd.notna(row['gig_sc_id']) and pd.notna(row['previous_order_sc_id']):
+        return int(row['gig_sc_id'] == row['previous_order_sc_id'])
+
+def compare_order_amounts(row):
+    if pd.notna(row['gig_price']) and pd.notna(row['previous_order_amount']) and row['previous_order_amount'] > 0:
+        return row['gig_price'] / row['previous_order_amount']
```
Comment (amardavid-cmd) [inline ctr_model.py:85]: return 0 (for the else)

File: ctr_model.py [hunk 8]
```diff
@@ -1 +1,156 @@
...
+for col in categorical_columns:
+    full_df[col] = rare_cats_to_other(full_df[col])
+    full_df[col].fillna('other', inplace=True)
+
+for col in numeric_columns:
+    full_df[col].fillna(0, inplace=True)
+
+def compare_sc_id(row):
+    if pd.notna(row['gig_sc_id']) and pd.notna(row['previous_order_sc_id']):
+        return int(row['gig_sc_id'] == row['previous_order_sc_id'])
```
Comment (amardavid-cmd) [inline ctr_model.py:81]: return 0 for the else or maybe -1 for identify it didn't stand in the condition, but because we already fullfill the null values for those categorial columns the condition is not needed, but another issue- there is no indicates if it is other , rare category so maybe we need to add -1 for the other , because if is is other on both of them it is still doesn't mean it from the same category

File: ctr_model.py [hunk 9]
```diff
@@ -1 +1,156 @@
...
+full_df = add_gig_features(full_df)
+
+x_train, x_test, y_train, y_test = train_test_split(full_df.drop('is_click', axis=1), full_df['is_click'], test_size=0.2)
+x_train, x_val, y_train, y_val = train_test_split(x_train, y_train, test_size=0.2)
+
+for row in full_df.dtypes.reset_index().iterrows():
+    if (row[1][0] == 'object'):
+        full_df[row[1]['index']] = full_df[row[1]['index']].astype('category')
+
+model_params = dict(
```
Comment (amardavid-cmd) [inline ctr_model.py:102]: on the model params- try to do hyper parameters tuning- change the scale_pos_weight because the data is unbalance we see it in the pie diagram before on the is_click column

File: ctr_model.py [hunk 10]
```diff
@@ -1 +1,156 @@
...
+    bins=1000
+    full_df[col].plot(kind='hist', bins=bins, figsize=(13, 5), color='b', alpha=0.5, density=True)
+    plt.title(f'{col} histogram')
+    plt.show()
+
+layout = np.arange(np.array(binary_columns).shape[0]).reshape(-1, 2)
+fig, axs = plt.subplot_mosaic(layout, figsize=(10, 10))
+for i, col in enumerate(binary_columns):
+    ser = full_df[col].value_counts()
+    ser.plot(kind='pie', autopct='%1.1f%%', title=col, ylabel='', ax=axs[i])
```
Comment (amardavid-cmd) [inline ctr_model.py:47]: can do it not in for loop

File: ctr_model.py [hunk 11]
```diff
@@ -1 +1,156 @@
...
+pd.DataFrame.from_dict(feature_importance).mean().sort_values().plot(kind='barh', figsize=(8, 10))
+plt.title('Mean feature importance (cover)')
+plt.show();
+
+means = cv_df.mean()
+stds = cv_df.std()
+plt.barh(means.index, means.values, xerr=stds.values, capsize=5, alpha=0.5)
+plt.title('Mean CV classification metrics')
+plt.ylabel('Score')
+plt.grid(axis='x');
```
Comment (amardavid-cmd) [inline ctr_model.py:156]: on the feature important all the text columns- search query, gig title and context didn't contribute to the model- we can classified those columns - use some transform in order to use them in the model

File: ctr_model.py [hunk 12]
```diff
@@ -1 +1,156 @@
...
+    df['is_weekend'] = df.apply(is_weekend, axis=1)
+    df = df.drop(['created_at','previous_order_date','user_reg_date'], axis=1)
+    return df
+
+full_df = add_time_features(full_df)
+
+def rare_cats_to_other(ser, thresh=0.05):
+    value_counts = ser.value_counts(normalize=True)
+    values_to_replace = value_counts[value_counts < thresh].index
+    return ser.replace(values_to_replace, 'other')
```
Comment (amardavid-cmd) [inline ctr_model.py:70]: all the text category will be other- for example the search query column

File: ctr_model.py (hunk 5 above)
Comment (amardavid-cmd) [reply ctr_model.py:74]: another issue - i don't sure all the numeric columns need the same default, for example ggi_price I think it should be fullfill with the avg price because there is a meanning if we put 0 there

Comment (amardavid-cmd) [reply ctr_model.py:74]: another issue we added columns but they aren't include in those first columns and there nan won't be treatment- for example in continent we have nan values

File: ctr_model.py [hunk 13]
```diff
@@ -1 +1,156 @@
...
+full_df['continent'] = full_df['user_timezone'].apply(lambda x: x.split('/')[0].lower())
+vc = full_df.continent.value_counts()
+vc.plot(kind='bar')
+
+temp = full_df.groupby('gig_sc_id').agg(
+    num_clicks=('is_click', 'sum'),
+    group_size=('gig_sc_id', 'size')
+)
+temp['sc_id_ctr'] = temp.num_clicks / temp.group_size
+full_df = full_df.join(temp, on='gig_sc_id').drop(['num_clicks','group_size'], axis=1)
```
Comment (amardavid-cmd) [inline ctr_model.py:35]: this is a problematic feature because it can cause a data leakage- it calculate on all the ds rows but we don't have this information in the test dataset. if we want to use this we need to calculate it after the split and add this as a fix data to the test df

File: ctr_model.py [hunk 14]
```diff
@@ -1 +1,156 @@
...
+    scale_pos_weight=2,
+)
+
+feature_importance = []
+cv_metrics = []
+n_folds = 6
+full_df = full_df.reset_index(drop=True)
+
+splitter = KFold(n_splits=n_folds, shuffle=True, random_state=13)
+for i, (train_indices, test_indices) in enumerate(splitter.split(full_df)):
```
Comment (amardavid-cmd) [inline ctr_model.py:119]: the split here is problamatic because it split from the all dataset but we shouldn't split the test ds each itteration, so we need to change the split here and in each fold just split the train and validation datasets

File: ctr_model.py (hunk 7 above)
Comment (amardavid-cmd) [reply ctr_model.py:85]: in this we assume the gig is on fix price but if the price of the gig can change - this is not right, because the current price is not the price in the previous order

Comment (amardavid-cmd) [reply ctr_model.py:85]: and we already fulfill the null values so we don't need to check it

File: ctr_model.py (hunk 11 above)
Comment (amardavid-cmd) [reply ctr_model.py:156]: another things I should do and I didn't see here:
in the first I will add a correlation matrix to see if the feature are depends each other, if there is such a strong correlation we should think to drop those feature.
I think that should be more feature I could think to add and maybe help with this model.
I would also look on the numeric distribution I think it is is important to look on and to see if it better to normalize all of them or part in order to use the same scale
and also for the end - use hyper petameters tuning, and check if other models give a better accuracy

Comment (amardavid-cmd) [reply ctr_model.py:156]: and also add the text to the model I think this can be a big influence on the model
